pkg/upgrade/upgrade.go

```diff
@@ -56,243 +63,183 @@ const CredentialsSecretName string = "weave-gitops-enterprise-credentials"
 
-	return toUpgradeConfigs(uv)
+	return err
 }
 
-func toUpgradeConfigs(uv UpgradeValues) (*UpgradeConfigs, error) {
-	gitRepoNamespace, gitRepoName, err := getGitRepositoryNamespaceAndName(uv.GitRepository)
-	if err != nil {
-		return nil, err
+func marshalToYamlStream(objects []runtime.Object) ([]byte, error) {
+	out := [][]byte{}
+
+	for _, obj := range objects {
+		b, err := yaml.Marshal(obj)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal HelmRepository object to YAML: %w", err)
+		}
+
+		out = append(out, b)
 	}
 
-	return &UpgradeConfigs{
-		CLIGitConfig: pctl_git.CLIGitConfig{
-			Directory: ".",
-			Branch:    uv.HeadBranch,
-			Remote:    uv.Remote,
-			Base:      uv.BaseBranch,
-			Message:   uv.CommitMessage,
-		},
-		SCMConfig: pctl_git.SCMConfig{
-			Branch: uv.HeadBranch,
-			Base:   uv.BaseBranch,
-			Repo:   uv.RepoOrgAndName,
+	return bytes.Join(out, []byte("---\n")), nil
+}
+
+func makeHelmResources(namespace, version, cname, repoURL string, values []string) ([]runtime.Object, error) {
+	helmRepository := &sourcev1.HelmRepository{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "weave-gitops-enterprise-charts",
+			Namespace: namespace,
 		},
-		InstallConfig: install.Config{
-			RootDir:          filepath.Join(uv.Out, "weave-gitops-enterprise"),
-			GitRepoNamespace: gitRepoNamespace,
-			GitRepoName:      gitRepoName,
+		TypeMeta: metav1.TypeMeta{
+			Kind:       sourcev1.HelmRepositoryKind,
+			APIVersion: sourcev1.GroupVersion.String(),
 		},
-		Profile: catalog.Profile{
-			ProfileConfig: catalog.ProfileConfig{
-				ConfigMap:     uv.ConfigMap,
-				Namespace:     uv.Namespace,
-				Path:          "weave-gitops-enterprise",
-				ProfileBranch: uv.ProfileBranch,
-				SubName:       "weave-gitops-enterprise",
-			},
-			GitRepoConfig: catalog.GitRepoConfig{
-				Namespace: gitRepoNamespace,
-				Name:      gitRepoName,
+		Spec: sourcev1.HelmRepositorySpec{
+			Interval: metav1.Duration{Duration: time.Minute},
+			URL:      EnterpriseChartURL,
+			SecretRef: &meta.LocalObjectReference{
+				Name: CredentialsSecretName,
 			},
 		},
-	}, nil
-}
-
-func upgrade(ctx context.Context, enterpriseProfileURL string, uc UpgradeConfigs, gitClient git.Git, pctlGitClient pctl_git.Git, pctlSCMClient pctl_git.SCMClient, dryRun bool) error {
-	tempDir, err := ioutil.TempDir("", "git-")
-	if err != nil {
-		return fmt.Errorf("failed to create temp folder for remote git clone of profiles repo: %w", err)
 	}
 
-	defer os.RemoveAll(tempDir)
-
-	_, err = gitClient.Clone(
-		ctx,
-		tempDir,
-		enterpriseProfileURL,
-		uc.Profile.ProfileConfig.ProfileBranch,
-	)
-	if err != nil {
-		return fmt.Errorf("failed to clone git repo: %v", err)
+	if version == "" {
+		version = "latest"
 	}
 
-	profileDefinition := uc.Profile
-	profileDefinition.URL = "file://" + tempDir
+	valuesTemplate := `{
+		"config": {
+			"cluster": {
+				"name": "%s"
+			},
+			"capi": {
+				"repositoryURL": "%s"
+			}
+		}
+	}`
```

Comment:
We can do a literal here to avoid the decode w/ something like this (made a little demo 🎉 )

https://play.golang.org/p/iFTT9sE9FIx

Reply:
Thank you very much Simon, that looks much better than using strings too 👍. I tried to do it with nested `map[string]interface{}` as well but it doesn't look very readable 😅